# Introduction {#introduction}

In this section we will introduce some key concepts and features of **scMINER2**.

## Motivation
The development of the **scMINER2** algorithm is motivated by the concept of learning *causal features* through integration of gene expression and chromatin accessibility data using network-based approach. Single-cell technology, such as scRNA and scATAC, has enabled data acquisition of multi-modal cellular states. However, due to inherent stochastic and sparse nature of such data generation process foreshadows any causal interpretation of the features identified through analysis of such data. Here, borrowing the concept of *Reinforcement Learning*, we design a computational approach to reconstruct a reinforced network from scRNA and scATAC data, followed by network wide search of network motifs (e.g. feed-forward loops) as an *attention* mechanism and reconstruct the **Rewiring Matrix**. Decomposition of this **Rewiring Matrix** allows the user to unbiasedly infer locally enriched sub graphs (TF-TF interactions) from the global reinforced network, which might help explain the input cell states as well as capture the underlying cellular differentiation trajectory. 

**scMINER2** algorithm consists of three key steps.

## Step 1: Network Encoding
It is a computational approach to link TFs to cis-regulatory elements in the form of scATAC peaks adapted from the study [@insilico] to reconstruct transcription factor (TF) regulatory network followed by greedy or context aware merging with TF regulatory network reconstructed from gene expression data using SJARACNe [@sjaracne]. In brief, scATAC-seq peaks that have high predicted binding of the transcription factor inferred through motif enrichment are selected and further refined using Pearson correlation between accessibility of the peak and the expression of the transcription factor predicted to bind at the peak using the formula: 
$$
X_{ij} \;=\; \sigma_{ij}\,\operatorname{minmax}\!\bigl(\theta_{ij}\,\pi_{i}\bigr)
$$
**\(X_{ij}\)** = TF-binding score of TF \(j\) in peak \(i\)\
**\(\sigma_{ij}\)** = Pearson correlation between the chromatin accessibility of peak \(i\) and RNA expression of TF \(j\)\
**\(\theta_{ij}\)** = Motif enrichment score for TF \(j\) in peak \(i\)\
**\(\pi_{ij}\)** = Mean chromatin accessibility for peak \(i\) across all samples\

## Step 2: Bottleneck
The key innovation of **scMINER2** algorithm is implemented in this step. To extract the network bottleneck, a network wide search and classification of network motifs is performed as (e.g. feed-forward loops) described in the study [@milo] and nodes are ranked utilizing a degree preserving randomization and empirical score calculation to rank the nodes after 1000 permutation of the reinforced network [@maslov]. Top nodes from this step are defined as *attention nodes* and peaks that have positive TF-binding score (**\(X_{ij}\)** \(> 0\) are defined as bottleneck edges. The chromatin accessibility of those edges are then summarized to reconstruct the **Rewiring Matrix**, **\(R_{ij}\)**, whose rows \(i\) represent the edges and columns \(j\) represent samples and values represent the *area under chromatin accessibility peak* or *(AUCAP)*. The values in the matrix **\(R_{ij}\)** ranges from \(0\) to \(+Inf\).

## Step 3: Decoding
In the decoding step, the reconstructed **Rewiring Matrix** **\(R_{ij}\)** is decomposed using consensus Non-Negative Matrix Factorization (cNMF) [@cnmf] to obtain locally enriched sub-graphs of the *attention nodes* that are specifically enriched in the input cell states. The reinforced network is then split into subset of networks called *hidden Gene Regulatory Networks* or *hGRN~1~*, *hGRN~2~*, *...*, *hGRN~n~* each of which is predicted to represent a cell state in the input. Further, this step supports extraction of edges specific to the cell state that might have been rewired as the cells are undergoing differentiation. This inference is further complemented by slingshot lineage tracing using **\(R_{ij}\)** to simultaneously infer the trajectory and underlying regulatory mechanism **(hGRNs)** of the cellular differentiation states.
